File: Aigilas/Aigilas/Creatures/CreatureFactory.cs
using System;
using Aigilas.Entities;
using Aigilas.Skills;
using SPX.Core;
using SPX.Entities;
using System.Collections.Generic;

namespace Aigilas.Creatures
{
    public class CreatureFactory
    {
        private static int __playerCount = 0;

        public static ICreature Create(int actorType, Point2 position)
        {
            AbstractCreature result;
            switch (actorType)
            {
                case AigilasActorType.PLAYER:
                    result = new Player(__playerCount++);
                    break;
                default:
                    result = GenerateCreature(actorType);
                    break;
            }
            result.Setup(position);
            EntityManager.AddObject(result);
            return result;
        }

        private static AbstractCreature GenerateCreature(int actorType)
        {
            switch(actorType)
            {
                case AigilasActorType.PEON:return new Peon();
                case AigilasActorType.ZORB:return new Zorb();
                case AigilasActorType.WRATH:return new Wrath();
                case AigilasActorType.HAND: return new Hand();
                case AigilasActorType.PRIDE: return new Pride();
                case AigilasActorType.ENVY: return new Envy();
                case AigilasActorType.GLUTTONY: return new Gluttony();
                case AigilasActorType.LUST: return new Lust();
                case AigilasActorType.SLOTH: return new Sloth();
                case AigilasActorType.GREED: return new Greed();
                case AigilasActorType.SERPENT: return new Serpent();
                case AigilasActorType.BREAKING_WHEEL: return new BreakingWheel();
                default:throw new Exception("No Factory generation logic for: "+actorType);
            }
        }

        public static ICreature CreateRandom(Point2 randomPoint)
        {
            var val = RNG.Rand.Next(0, Generate.Randoms.Count);
            return Create(Generate.Randoms[val], randomPoint);
        }

        public static ICreature CreateMinion(string skillId, ICreature source,SkillEffect effectGraphic=null,Point2 location=null)
        {
            Minion result = null;
            switch (skillId)
            {
                case SkillId.ACID_NOZZLE:result = new AcidNozzle();break;
                case SkillId.DART_TRAP: result = new DartTrap(); break;
                case SkillId.EXPLODE: result = new Explosion(); break;
                case SkillId.ICE_SHARD: result = new IceShard(); break; 
                case SkillId.VAPOR_CLOUD: result = new VaporCloud(); break;
                case SkillId.PLAGUE: result = new PoisonCloud(); break;
                default:
                    throw new Exception("No minion was defined for the given skillId.");
            }
            result.Init(source, effectGraphic);
            if (location != null)
            {
                result.SetLocation(location);
            }
            EntityManager.AddObject(result);
            return result;
        }

        private static List<int> __remainingBosses = new List<int>()
        {
            AigilasActorType.WRATH,
            AigilasActorType.ENVY,
            AigilasActorType.PRIDE,
            AigilasActorType.SLOTH,
            AigilasActorType.GREED,
            AigilasActorType.LUST,
            AigilasActorType.GLUTTONY
        };

        public static int BossesRemaining()
        {
            return __remainingBosses.Count;
        }

        public static IEntity CreateNextBoss(Point2 randomPoint)
        {
            int nextBoss = __remainingBosses[RNG.Rand.Next(0,__remainingBosses.Count)];
            __remainingBosses.Remove(nextBoss);
            return Create(nextBoss,randomPoint);
        }
    }
}

File: proj/src/aigilas/creatures/CreatureFactory.java
package aigilas.creatures;

import aigilas.entities.SkillEffect;
import aigilas.items.ItemFactory;
import aigilas.skills.SkillId;
import spx.bridge.ActorType;
import spx.core.Point2;
import spx.core.RNG;
import spx.entities.EntityManager;
import spx.entities.IEntity;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class CreatureFactory {
    private static int __playerCount = 0;

    public static BaseCreature create(ActorType actorType, Point2 position) {
        BaseEnemy result;
        switch (actorType) {
            case PLAYER:
                result = new Player(__playerCount++);
                break;
            default:
                result = generateCreature(actorType);
                break;
        }
        result.setup(position);
        EntityManager.addObject(result);
        // $$$ Testing inv HUD
        for (int ii = 0; ii < 100; ii++) {
            result.pickupItem(ItemFactory.createRandomPlain());
        }
        return result;
    }

    private static BaseEnemy generateCreature(ActorType actorType) {
        switch (actorType) {
            case PEON:
                return new Peon();
            case ZORB:
                return new Zorb();
            case WRATH:
                return new Wrath();
            case HAND:
                return new Hand();
            case PRIDE:
                return new Pride();
            case ENVY:
                return new Envy();
            case GLUTTONY:
                return new Gluttony();
            case LUST:
                return new Lust();
            case SLOTH:
                return new Sloth();
            case GREED:
                return new Greed();
            case SERPENT:
                return new Serpent();
            case BREAKING_WHEEL:
                return new BreakingWheel();
            default:
                try {
                    throw new Exception("No Factory generation logic for: " + actorType);
                } catch (Exception e) {

                    e.printStackTrace();
                }
                return null;
        }
    }

    public static BaseCreature createRandom(Point2 randomPoint) {
        return create(Generate.Randoms.get(RNG.next(0, Generate.Randoms.size())), randomPoint);
    }

    public static BaseCreature createMinion(SkillId skillId, BaseCreature source, SkillEffect effectGraphic, Point2 location) {
        Minion result = null;
        switch (skillId) {
            case ACID_NOZZLE:
                result = new AcidNozzle();
                break;
            case DART_TRAP:
                result = new DartTrap();
                break;
            case EXPLODE:
                result = new Explosion();
                break;
            case ICE_SHARD:
                result = new IceShard();
                break;
            case VAPOR_CLOUD:
                result = new VaporCloud();
                break;
            case PLAGUE:
                result = new PoisonCloud();
                break;
            default:
                try {
                    throw new Exception("No minion was defined for the given skillId.");
                } catch (Exception e) {

                    e.printStackTrace();
                }
        }
        result.init(source, effectGraphic);
        if (location != null) {
            result.setLocation(location);
        }
        EntityManager.addObject(result);
        return result;
    }

    public static BaseCreature createMinion(SkillId skillId, BaseCreature source, SkillEffect effectGraphic) {
        return createMinion(skillId, source, effectGraphic, null);
    }

    public static BaseCreature createMinion(SkillId skillId, BaseCreature source) {
        return createMinion(skillId, source, null, null);
    }

    private static List<ActorType> __remainingBosses;

    static {
        __remainingBosses = new LinkedList<>(Arrays.asList(ActorType.WRATH, ActorType.ENVY, ActorType.PRIDE, ActorType.SLOTH, ActorType.GREED, ActorType.LUST, ActorType.GLUTTONY));
    }

    public static void reset(){
        __remainingBosses = new LinkedList<>(Arrays.asList(ActorType.WRATH, ActorType.ENVY, ActorType.PRIDE, ActorType.SLOTH, ActorType.GREED, ActorType.LUST, ActorType.GLUTTONY));
        __playerCount = 0;
    }

    public static int bossesRemaining() {
        return __remainingBosses.size();
    }

    public static IEntity createNextBoss(Point2 randomPoint) {
        ActorType nextBoss = __remainingBosses.get(RNG.next(0, __remainingBosses.size()));
        __remainingBosses.remove(nextBoss);
        return create(nextBoss, randomPoint);
    }
}
